### PR TITLE
introduce sequence cache poisoning

### LINKED ISSUE #2231: A write followed by another write can very easily get aborted.
Take a look at this test

https://github.com/cockroachdb/cockroach/blob/master/sql/driver/driver_test.go#L236

The second transaction aborts the first one. Is this because it came with a higher priority? or did some retry logic get triggered that made it retry with a higher priority? If it did retry then the behavior is probably okay.

### PR DESCRIPTION
**Pull request by tbg**
on `ResolveIntent{,Range}`, atomically add sequence cache
entries for the pushed transaction which cause it to restart
or abort appropriately upon its next contact with the Range
during the current epoch.

fixes #2231.

[Review on Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/3157)

### CODE REVIEW

File: storage/replica.go [hunk 1]
```diff
@@ -1031,16 +1031,26 @@ func (r *Replica) applyRaftCommandInBatch(ctx context.Context, index uint64, ori
...
+		} else if ba.Txn.Epoch > epoch {
+			// No cache hit from current or future epoch, continue.
+		} else if sequence == roachpb.SequencePoisonAbort {
+			// We were poisoned, which means that our Transaction has been
+			// aborted and learns about that right now.
+			abortErr := roachpb.NewTransactionAbortedError(ba.Txn)
+			abortErr.Txn.Timestamp.Forward(entry.Timestamp)
+			return btch, nil, nil, abortErr
 		} else if sequence >= ba.Txn.Sequence || epoch > ba.Txn.Epoch {
 			// We hit the cache, so let the transaction restart.
```
Comment (bdarnell) [inline storage/replica.go]: Add a comment that `roachpb.SequencePoisonRestart` hits this path.

**Comment by bdarnell** [comment]
LGTM

**Comment by tamird** [comment]
Reviewed 7 of 7 files at r1, 4 of 4 files at r2, 6 of 6 files at r3.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

_[storage/sequence_cache.go, line 152 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3157#-K3MX34HOMgohn73SdI8:-K3MX34HOMgohn73SdI9:879411099) ([raw file](https://github.com/cockroachdb/cockroach/blob/ea1f8cfb0278ff643b9912280474974b95449c9b/storage/sequence_cache.go#L152)):_
this is in the wrong commit

---


_Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3157)_

**Comment by mrtracy** [comment]
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

_[storage/replica_test.go, line 2221 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3157#-K3MhfAeHI-Tye-TQoqp:-K3MhfAeHI-Tye-TQoqq:-744364828) ([raw file](https://github.com/cockroachdb/cockroach/blob/ea1f8cfb0278ff643b9912280474974b95449c9b/storage/replica_test.go#L2221)):_
In #2231, the core issues seems to be that "Read your own writes" is violated.  However, this test doesn't seem to address that, because we only assert that the pushee can't increment after the poisoning; can we assert that the pushee can't *read* a value after being pushed?

---


_Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3157)_

**Comment by mrtracy** [comment]
Made a suggestion in terms of the test, but LGTM otherwise.

---

Review status: all files reviewed at latest revision, 3 unresolved discussions.

---



_Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3157#-:-K3Mk3T7jyFRubVYLJQO:1831359558)_

File: storage/replica.go (hunk 1 above)
Comment (tschottdorf) [inline storage/replica.go]: done.

**Comment by tschottdorf** [comment]
Review status: 10 of 12 files reviewed at latest revision, 2 unresolved discussions.

---

_[storage/replica_test.go, line 2221 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3157#-K3MhfAeHI-Tye-TQoqp:-K3Q4yyqFH0jDobRHtBz:-949732036) ([raw file](https://github.com/cockroachdb/cockroach/blob/ea1f8cfb0278ff643b9912280474974b95449c9b/storage/replica_test.go#L2221)):_
You're right, that's an oversight. It's in the RFC, but forgot about it here. Added in new commit, PTAL.

---


_Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3157)_

**Comment by tschottdorf** [comment]
still got a race to fix, but then we should be good to go.

**Comment by tamird** [comment]
Reviewed 4 of 4 files at r4.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

_[storage/replica.go, line 773 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3157#-K3Qlamo2Z0q78uJWjsI:-K3Qlamo2Z0q78uJWjsJ:-1026318688) ([raw file](https://github.com/cockroachdb/cockroach/blob/6ce3380ac4adfff5fbb99fef6f4bbabd658fa0de/storage/replica.go#L773)):_
seems like this could be below the `err != nil` check below and then not need to overwrite `err` (use a new binding)

---


_Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3157)_

**Comment by tschottdorf** [comment]
Review status: 11 of 12 files reviewed at latest revision, 3 unresolved discussions.

---

_[storage/replica.go, line 773 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3157#-K3Qlamo2Z0q78uJWjsI:-K3R5-CJsvUDY4fkPpcR:-966768317) ([raw file](https://github.com/cockroachdb/cockroach/blob/6ce3380ac4adfff5fbb99fef6f4bbabd658fa0de/storage/replica.go#L773)):_
had to move this up a bit (before `endCmds`). Either the original (if any) or sequence cache error need to be passed to `endCmds`, so avoiding the assignment is going to look awkward.

---


_Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3157)_

**Comment by tamird** [comment]
Reviewed 1 of 1 files at r5.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



_Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3157)_

**Pull request merged**